src/crypto/generators.cpp

```diff
@@ -0,0 +1,192 @@
+static ge_p3 G_p3;
+static ge_p3 H_p3;
+static ge_cached G_cached;
+static ge_cached H_cached;
+
+// misc
+static std::once_flag init_gens_once_flag;
+
+//-------------------------------------------------------------------------------------------------------------------
+// hash-to-point: H_p(x) = 8*point_from_bytes(keccak(x))
+//-------------------------------------------------------------------------------------------------------------------
+static void hash_to_point(const hash &x, crypto::ec_point &point_out)
+{
+    hash h;
+    ge_p3 temp_p3;
+    ge_p2 temp_p2;
+    ge_p1p1 temp_p1p1;
+
+    cn_fast_hash(reinterpret_cast<const unsigned char*>(&x), sizeof(hash), h);
+    ge_fromfe_frombytes_vartime(&temp_p2, reinterpret_cast<const unsigned char*>(&h));
+    ge_mul8(&temp_p1p1, &temp_p2);
+    ge_p1p1_to_p3(&temp_p3, &temp_p1p1);
+    ge_p3_tobytes(to_bytes(point_out), &temp_p3);
+}
+//-------------------------------------------------------------------------------------------------------------------
+//-------------------------------------------------------------------------------------------------------------------
+static public_key reproduce_generator_G()
+{
+    // G = {x, 4/5 mod q}
+    fe four, five, inv_five, y;
+    fe_0(four);
+    fe_0(five);
+    four[0] = 4;
+    five[0] = 5;
+    fe_invert(inv_five, five);
+    fe_mul(y, four, inv_five);
+
+    public_key reproduced_G;
+    fe_tobytes(to_bytes(reproduced_G), y);
+
+    return reproduced_G;
+}
+//-------------------------------------------------------------------------------------------------------------------
+//-------------------------------------------------------------------------------------------------------------------
+static public_key reproduce_generator_H()
+{
+    // H = 8*to_point(keccak(G))
+    // note: this does not use the point_from_bytes() function found in H_p(), instead directly interpreting the
+    //       input bytes as a compressed point (this can fail, so should not be used generically)
+    // note2: to_point(keccak(G)) is known to succeed for the canonical value of G (it will fail 7/8ths of the time
+    //        normally)
+    ge_p3 temp_p3;
+    ge_p2 temp_p2;
+    ge_p1p1 temp_p1p1;
+
+    hash H_temp_hash{cn_fast_hash(to_bytes(G), sizeof(ec_point))};
+    (void)H_temp_hash;  //suppress unused warning
+    assert(ge_frombytes_vartime(&temp_p3, reinterpret_cast<const unsigned char*>(&H_temp_hash)) == 0);
+    ge_p3_to_p2(&temp_p2, &temp_p3);
+    ge_mul8(&temp_p1p1, &temp_p2);
+    ge_p1p1_to_p3(&temp_p3, &temp_p1p1);
+
+    public_key reproduced_H;
+    ge_p3_tobytes(to_bytes(reproduced_H), &temp_p3);
+
+    return reproduced_H;
+}
+//-------------------------------------------------------------------------------------------------------------------
+// Make generators, but only once
+//-------------------------------------------------------------------------------------------------------------------
+static void init_gens()
```

Comment:
I think it's really cool how you're showing your work through code; I genuinely think its refreshing. But since all the calls `get_G, getH, etc` call `init_gens` internally, this introduces extra release binary size and slows down runtime. Maybe this would be better as a unit test? Also, that way, it would be checked at compile-time on the many platforms we run CI on every single PR.

Reply:
The `init_gens()` is necessary for converting the `unsigned char[]` representation to `char[]`, which is what `crypto::public_key` has internally.

Reply:
You can have `G` and `H` represented directly as a `constexpr` if you put in the correct `char` values, so you don't need any copying at the beginning of the program, no initialization functions calls, or `std::once_flag`s. It's a little uglier to read because of the negative signs, but they're either the right values or they're not.

Reply:
Example:
```
constexpr const public_key G = { {88, 102, 102, 102, 102, 102, 102, 102, 102, 102, 102, 102, 102, 102, 102, 102, 102, 102, 102, 102, 102, 102, 102, 102, 102, 102, 102, 102, 102, 102, 102, 102} };
constexpr const public_key H = { {-117, 101, 89, 112, 21, 55, -103, -81, 42, -22, -36, -97, -15, -83, -48, -22, 108, 114, 81, -43, 65, 84, -49, -87, 44, 23, 58, 13, -45, -100, 31, -108} };
```

Reply:
Well you'd still need the once_flag to generate the other representations.

Reply:
Unless there's something I'm not seeing, you wouldn't need the whole `std::call_once` block since `G`, `H`, `G_p3`, `H_p3`, `G_cached`, and `H_cached` all have POD types and the process for deriving these constants are deterministic. It's always going to be the same in the end so you can just cut out that work right? If you still want to check the constants at runtime on every run for debug builds, we can still do this if we shuffle everything into a function `reproduce_and_check_all_generators` or something. 

We can also have the raw byte representations in a comment right above it as well as the mathematical way to derive the point.

Reply:
That way there's 0 overhead but we still get the same safety guarantees.

Reply:
Okay I did better performance tests and dang... the compiler was better at optimizing that stuff away than I was expecting. Once that dependency is resolved, I think I'll be good.

Reply:
Which dependency?

Reply:
Adding more to the discussion - the compiler is good at optimizing, but every `init_gens()` call produces a read-acquire statement that requires synchronization across cores. As an example, use gcc godbolt to generate code for a function-local static - the x86-64 ASM looks like a simple double-check lock but the arm64 ASM has a `dmb` instruction (memory barrier). This isn't a huge impact on performance, but does add stalls/wait cycles.

Constexpr was expanded in C++14, so compile-time conversion and 2's complement checking should now be possible across platforms:
```c++
template<typename T>
constexpr T bytes_to(const std::initializer_list<unsigned char> in)
{
    T out{}; // zero-initialize trailing bytes

    auto current = std::begin(out.data);
    if (std::end(out.data) - current < in.size())
        throw std::runtime_error{"invalid initializer size"};

    for (const auto byte : in)
        *current++ = byte;
    return out;
}

struct char_test { char data[5]; };
constexpr const auto test_case = bytes_to<char_test>({0x00, 0x7f, 0x80, 0xff});
static_assert(test_case.data[0] == 0, "");
static_assert(test_case.data[1] == 127, "");
static_assert(test_case.data[2] == char(-128), "");
static_assert(test_case.data[3] == char(-1), "");
static_assert(test_case.data[4] == 0, "");

constexpr const auto G = bytes_to<public_key>(
    {0x58, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 
    0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66, 0x66}
);
```
Tested against x86-64, arm64, and risc-v on gcc godbolt.

Reply:
And I was suggesting this for `G` and `H`, but leaving the `init_gens()` for the other values. The test harness will still call `init_gens()` by calling values _other than_ `G` and `H`.